editoast/src/views/timetable/similar_trains/trains_traffic.rs

```diff
@@ -0,0 +1,307 @@
+            let trains_for_speedlimit = self
+                .trains_by_speedlimit
+                .entry(train_data.speed_limit_tag)
+                .or_default();
+            trains_for_speedlimit.insert(train_data.id);
+
+            // Index the train by its rollingstock
+            let trains_for_rollingstock = self
+                .trains_by_rollingstock
+                .entry(train_data.rolling_stock.clone())
+                .or_default();
+            trains_for_rollingstock.insert(train_data.id);
+
+            // Index the train by segment
+            for (from, to) in train_data
+                .operational_points
+                .iter()
+                .filter(|op| op.stop)
+                .tuple_windows()
+            {
+                let trains_for_segment = self
+                    .trains_by_segment
+                    .entry(Segment {
+                        from: from.clone(),
+                        to: to.clone(),
+                    })
+                    .or_default();
+                trains_for_segment.insert(train_data.id);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Find in the traffic a composition of trains that match the given train itinerary
+    pub fn find_compatible_trains(
+        &self,
+        rolling_stock: Option<String>,
+        speed_limit_tag: Option<String>,
+        waypoints: Vec<Waypoint>,
+    ) -> Vec<TrainTraffic> {
+        let mut valid_trains = HashSet::<usize>::new();
+
+        // Getting a list of valid train Ids which match the train's specifications
+        let compliant_trains = match (rolling_stock, speed_limit_tag) {
+            (Some(rs), None) => self
+                .trains_by_rollingstock
+                .get(&rs)
+                .cloned()
+                .unwrap_or_default(),
+            (None, Some(speed)) => self
+                .trains_by_speedlimit
+                .get(&speed)
+                .cloned()
+                .unwrap_or_default(),
+            (Some(rs), Some(speed)) => {
+                let rs_trains = self
+                    .trains_by_rollingstock
+                    .get(&rs)
+                    .cloned()
+                    .unwrap_or_default();
+                let speed_trains = self
+                    .trains_by_speedlimit
+                    .get(&speed)
+                    .cloned()
+                    .unwrap_or_default();
+                rs_trains.intersection(&speed_trains).cloned().collect()
+            }
+            _ => self.trains_by_id.keys().cloned().collect(),
+        };
+
+        tracing::debug!(
+            nb_train = compliant_trains.len(),
+            "Compliant trains by specifications"
+        );
+
+        for (from, to) in waypoints.into_iter().filter(|w| w.stop).tuple_windows() {
+            let trains_in_segment = self.trains_by_segment.get(&Segment { from, to });
+            if let Some(ids) = trains_in_segment {
+                for index in ids.iter() {
```

Comment:
Here is an equivalent form of writing this

```suggestion
                for index in &ids {
```